File: packages/frontend/types/task.ts
export enum TaskPriority {
  LOW = "low",
  MEDIUM = "medium",
  HIGH = "high",
  URGENT = "urgent",
}

export enum TaskStatus {
  TODO = "todo",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
}

export interface Task {
  id: number;
  title: string;
  description?: string;
  priority: TaskPriority;
  status: TaskStatus;
  due_date?: string | null; // ISO string
  position?: number | null;
  tags: string[];
  goal_id?: number | null;
  created_at: string;
  updated_at: string;
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority?: TaskPriority;
  status?: TaskStatus;
  due_date?: string;
  position?: number | null;
  tags?: string[];
  goal_id?: number;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: TaskPriority;
  status?: TaskStatus;
  due_date?: string | null; // Allow clearing date
  position?: number | null;
  tags?: string[];
  goal_id?: number | null;
}

export interface TaskFilters {
  status?: TaskStatus;
  priority?: TaskPriority;
  goal_id?: number;
  tag?: string;
  search?: string;
}

export interface TaskListResponse {
  items: Task[];
  count: number;
}

File: packages/backend/app/models/task.py
import enum
from datetime import datetime
from typing import List, Optional

from sqlalchemy import Column, Integer, String, Text, Enum, DateTime, JSON, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base


class TaskPriority(str, enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class TaskStatus(str, enum.Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"


class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    priority = Column(Enum(TaskPriority), default=TaskPriority.MEDIUM, nullable=False)
    status = Column(Enum(TaskStatus), default=TaskStatus.TODO, nullable=False, index=True)
    due_date = Column(DateTime, nullable=True, index=True)
    position = Column(Integer, nullable=True, index=True)
    tags = Column(JSON, default=list, nullable=False)
    
    # Foreign key to Goal
    goal_id = Column(Integer, ForeignKey("goals.id"), nullable=True, index=True)
    goal = relationship("Goal", backref="tasks")
    
    created_at = Column(DateTime, server_default=func.now(), nullable=False)
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False)

    def __repr__(self):
        return f"<Task(id={self.id}, title='{self.title}', status='{self.status}')>"

